### PR TITLE
[OM-96680] create container vertical scale CRD in turbo-policy

### PR DESCRIPTION
**Pull request by libai98**

### CODE REVIEW

File: api/v1alpha1/containercerticalscaleSettings_types.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+// LimitResourceConstraints defines the resize constraint for resource like CPU limit or Memory limit.
+type LimitResourceConstraint struct {
+	Max               *string `json:"max,omitempty"`
+	Min               *string `json:"min,omitempty"`
+	RecommendAboveMax *bool   `json:"recommendAboveMax,omitempty"`
+	RecommendBelowMin *bool   `json:"recommendBelowMin,omitempty"`
+}
+
+// RequestResourceConstraint defines the resize constraint for resource like CPU request or Memory request
```
Comment (ading1977) [inline api/v1alpha1/containercerticalscaleSettings_types.go:77]: Recommend adding a comment saying that we only resize down request for now.

File: api/v1alpha1/containercerticalscaleSettings_types.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+// The sample period ensures historical data for a minimu numbers of days before calculationg Agressiveness.
+// This ensures a minimum set of data points before the action is generated.
+type MinObservationPeriod string
+
+const (
+	None      MinObservationPeriod = "None"
+	OneDay    MinObservationPeriod = "1 Day"
+	ThreeDays MinObservationPeriod = "30 Days"
+	SevenDays MinObservationPeriod = "7 Days"
```
Comment (ading1977) [inline api/v1alpha1/containercerticalscaleSettings_types.go:36]: `30 Days` -> `3 Days`

Will it be better to use:`1D`, `3D`, `7D` (case insensitive?). This way, user doesn't need to add quotation marks.

File: api/v1alpha1/containercerticalscaleSettings_types.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+)
+
+type MaxObservationPeriod string
+
+const (
+	Last90Days MaxObservationPeriod = "Last 90 Days"
+	Last30Days MaxObservationPeriod = "1 Day"
+	Last7Days  MaxObservationPeriod = "Last 7 Days"
```
Comment (ading1977) [inline api/v1alpha1/containercerticalscaleSettings_types.go:44]: `1 Day` -> `Last 30 Days`?

Same as before, can we use `90D`, `30D`, `7D`?

File: api/v1alpha1/containercerticalscaleSettings_types.go (hunk 2 above)
Comment (libai98) [reply api/v1alpha1/containercerticalscaleSettings_types.go:36]: I like the idea of use "1D" and "1d" without quotation but am not sure in go how we can have same constant have two values defined.

Comment (libai98) [reply api/v1alpha1/containercerticalscaleSettings_types.go:36]: Maybe I defined all constant with lower case values and then write a function when we convert the input to ignore the case.

Comment (ading1977) [reply api/v1alpha1/containercerticalscaleSettings_types.go:36]: That sounds fine to me. At least we should support lower case.

File: config/samples/policy_v1alpha1_containerverticalscale.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    observationPeriod:
+      min: None
+      max: None
+
+    rateOfResize: High
+    agreesiveness: P99
+
+  behavior:
+    resize: manual
```
[the remark this answers is not in the document]
Comment (libai98) [reply config/samples/policy_v1alpha1_containerverticalscale.yaml:44]: Not quite get this one.Should I remove the new line in line 42 or add new line between 43 and 44?

[the remark this answers is not in the document]
Comment (libai98) [reply config/samples/policy_v1alpha1_containerverticalscale.yaml:44]: Done. Thanks!

**Review by ading1977 — APPROVED** [review]

**Comment by libai98** [comment]
These are modified files:
       api/v1alpha1/policy_constant.go
       api/v1alpha1/containercerticalscaleSettings_types.go
       api/v1alpha1/containerresourceconstraint_types.go
       api/v1alpha1/containerverticalscale_types.go
       api/v1alpha1/policy_target_reference.go
      config/samples/policy_v1alpha1_containerverticalscale.yaml

with generated CRD in 
      config/crd/bases/policy.turbonomic.io_containerverticalscales.yaml

File: api/v1alpha1/policy_settings.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+package v1alpha1
+
+import v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+const (
+	ResponseTime = "ResponseTime"
+	Transaction  = "Transaction"
+)
+
+type SloHorizontalScalePolicySetting struct {
```
Comment (ading1977) [inline api/v1alpha1/policy_settings.go:26]: nit: `SloHorizontalScalePolicySetting` -> `SLOHorizontalScalePolicySetting

**Review by ading1977 — APPROVED** [review]
Just a small nit.
Otherwise looks good to me.

File: api/v1alpha1/policy_settings.go (hunk 5 above)
Comment (libai98) [reply api/v1alpha1/policy_settings.go]: fixed. Thanks!

**Pull request merged**